File: plugins/datasources/yesoreyeram-directus-datasource/pkg/plugin/filter.go
package plugin

import (
	"strings"
)

// FilterNode is a node in the structured filter tree sent from the query editor.
// A node is either a condition (kind="condition") or a group (kind="group").
//
// Conditions compile to {"field": {"_op": value}}; groups compile to
// {"_and": [...]} / {"_or": [...]}. See
// https://docs.directus.io/guides/connect/filter-rules
type FilterNode struct {
	Kind string `json:"kind"`

	// Condition fields.
	Field string `json:"field,omitempty"`
	Op    string `json:"op,omitempty"`
	Value string `json:"value,omitempty"`

	// Group fields.
	Connector string       `json:"connector,omitempty"`
	Children  []FilterNode `json:"children,omitempty"`
}

// SortItem is a single structured sort directive persisted by the query editor.
type SortItem struct {
	Field     string `json:"field"`
	Direction string `json:"direction"`
}

// operatorMap maps the query editor's operator keys to Directus filter
// operators. The `not_empty` key is retained as a backward-compatible alias for
// previously-saved queries.
var operatorMap = map[string]string{
	"eq":          "_eq",
	"neq":         "_neq",
	"lt":          "_lt",
	"lte":         "_lte",
	"gt":          "_gt",
	"gte":         "_gte",
	"contains":    "_contains",
	"ncontains":   "_ncontains",
	"icontains":   "_icontains",
	"nicontains":  "_nicontains",
	"startsWith":  "_starts_with",
	"nstartsWith": "_nstarts_with",
	"endsWith":    "_ends_with",
	"nendsWith":   "_nends_with",
	"in":          "_in",
	"nin":         "_nin",
	"between":     "_between",
	"nbetween":    "_nbetween",
	"null":        "_null",
	"nnull":       "_nnull",
	"empty":       "_empty",
	"nempty":      "_nempty",
	"not_empty":   "_nempty",
}

// operatorArity classifies how an operator's value is interpreted:
//
//	"none"    – unary (encoded as boolean true), e.g. _null/_empty
//	"list"    – comma-separated tokens compiled into a JSON array, e.g. _in
//	"between" – exactly two comma-separated tokens compiled into a [min,max] array
//	"single"  – a single scalar value (default)
func operatorArity(op string) string {
	switch op {
	case "null", "nnull", "empty", "nempty", "not_empty":
		return "none"
	case "in", "nin":
		return "list"
	case "between", "nbetween":
		return "between"
	default:
		return "single"
	}
}

// BuildFilter converts a structured filter tree (the root group) into a Directus
// JSON filter object. It returns nil when there are no valid conditions, so the
// caller can omit the filter entirely.
//
// Directus filter format:
//
//	{"field_name": {"_eq": "value"}}
//	{"_and": [{"field": {"_eq": "val1"}}, {"field2": {"_neq": "val2"}}]}
func BuildFilter(root *FilterNode) map[string]any {
	if root == nil {
		return nil
	}
	return buildGroupFilter(root)
}

func buildGroupFilter(group *FilterNode) map[string]any {
	parts := make([]map[string]any, 0, len(group.Children))
	for i := range group.Children {
		child := group.Children[i]
		var frag map[string]any
		if child.Kind == "group" {
			frag = buildGroupFilter(&child)
		} else {
			frag = buildConditionFilter(child)
		}
		if frag != nil {
			parts = append(parts, frag)
		}
	}

	if len(parts) == 0 {
		return nil
	}
	if len(parts) == 1 {
		return parts[0]
	}

	key := "_and"
	if strings.EqualFold(group.Connector, "or") {
		key = "_or"
	}
	return map[string]any{key: parts}
}

func buildConditionFilter(c FilterNode) map[string]any {
	field := strings.TrimSpace(c.Field)
	if field == "" {
		return nil
	}
	op := strings.TrimSpace(c.Op)
	if op == "" {
		op = "eq"
	}

	directusOp, ok := operatorMap[op]
	if !ok {
		directusOp = "_eq"
	}

	var value any
	switch operatorArity(op) {
	case "none":
		// Directus expects a boolean true for null/empty checks.
		value = true
	case "list":
		tokens := parseListValue(c.Value)
		if len(tokens) == 0 {
			return nil
		}
		value = tokens
	case "between":
		tokens := parseListValue(c.Value)
		if len(tokens) != 2 {
			// _between requires exactly two values; drop incomplete ranges.
			return nil
		}
		value = tokens
	default: // single
		v := strings.TrimSpace(c.Value)
		if v == "" {
			// Drop incomplete conditions so an empty value is not sent as a
			// filter for the empty string.
			return nil
		}
		value = v
	}

	return map[string]any{
		field: map[string]any{
			directusOp: value,
		},
	}
}

// parseListValue splits a comma-separated value into trimmed, non-empty tokens.
func parseListValue(s string) []string {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

File: plugins/datasources/yesoreyeram-directus-datasource/src/filter.ts
import { SelectableValue } from '@grafana/data';

// ---------------------------------------------------------------------------
// Filter model
//
// The structured filter tree is compiled into a Directus JSON filter object on
// the backend. The frontend only describes conditions/groups; it does not build
// the filter itself.
// ---------------------------------------------------------------------------

export type LogicalConnector = 'and' | 'or';

export interface FilterCondition {
  kind: 'condition';
  /** Directus field name. */
  field: string;
  /** Operator (see OPERATORS). */
  op: string;
  /** Comparison value. Ignored for unary operators. */
  value?: string;
}

export interface FilterGroup {
  kind: 'group';
  /** How the direct children are combined (AND/OR). */
  connector: LogicalConnector;
  children: FilterNode[];
}

export type FilterNode = FilterCondition | FilterGroup;

export function newCondition(): FilterCondition {
  return { kind: 'condition', field: '', op: 'eq', value: '' };
}

export function newGroup(): FilterGroup {
  return { kind: 'group', connector: 'and', children: [] };
}

export function emptyRootGroup(): FilterGroup {
  return newGroup();
}

// ---------------------------------------------------------------------------
// Operator catalog (field-type aware)
// ---------------------------------------------------------------------------

export interface OperatorDef {
  value: string;
  label: string;
  /**
   * 'none'   = unary (no value)
   * 'single' = one value
   */
  arity: 'none' | 'single';
}

const EQUALITY: OperatorDef[] = [
  { value: 'eq', label: '=', arity: 'single' },
  { value: 'neq', label: '!=', arity: 'single' },
];

const NULLABLE: OperatorDef[] = [
  { value: 'empty', label: 'is empty', arity: 'none' },
  { value: 'not_empty', label: 'is not empty', arity: 'none' },
];

const COMPARISON: OperatorDef[] = [
  { value: 'gt', label: '>', arity: 'single' },
  { value: 'gte', label: '>=', arity: 'single' },
  { value: 'lt', label: '<', arity: 'single' },
  { value: 'lte', label: '<=', arity: 'single' },
];

const TEXT_MATCH: OperatorDef[] = [
  { value: 'contains', label: 'contains', arity: 'single' },
  { value: 'ncontains', label: 'does not contain', arity: 'single' },
  { value: 'startsWith', label: 'starts with', arity: 'single' },
  { value: 'endsWith', label: 'ends with', arity: 'single' },
];

const LIST_MATCH: OperatorDef[] = [
  { value: 'in', label: 'is one of (csv)', arity: 'single' },
  { value: 'nin', label: 'is not one of (csv)', arity: 'single' },
];

const RANGE: OperatorDef[] = [
  { value: 'between', label: 'is between (min,max)', arity: 'single' },
  { value: 'nbetween', label: 'is not between (min,max)', arity: 'single' },
];

/** Logical categories a Directus field type can map to for operator selection. */
export type FieldCategory = 'text' | 'number' | 'boolean' | 'date' | 'json';

const TYPE_CATEGORY: Record<string, FieldCategory> = {
  string: 'text',
  text: 'text',
  varchar: 'text',
  char: 'text',
  tinytext: 'text',
  mediumtext: 'text',
  longtext: 'text',
  json: 'json',
  alias: 'text',
  integer: 'number',
  bigInteger: 'number',
  smallInteger: 'number',
  decimal: 'number',
  float: 'number',
  double: 'number',
  boolean: 'boolean',
  date: 'date',
  dateTime: 'date',
  timestamp: 'date',
  time: 'date',
};

export function categoryForType(type?: string): FieldCategory {
  if (!type) {
    return 'text';
  }
  return TYPE_CATEGORY[type] ?? 'text';
}

export function operatorsForType(type?: string): OperatorDef[] {
  switch (categoryForType(type)) {
    case 'number':
      return [...EQUALITY, ...COMPARISON, ...RANGE, ...LIST_MATCH, ...NULLABLE];
    case 'boolean':
      return [...EQUALITY, ...NULLABLE];
    case 'date':
      return [...EQUALITY, ...COMPARISON, ...RANGE, ...NULLABLE];
    case 'json':
      return [...EQUALITY, ...NULLABLE];
    case 'text':
    default:
      return [...EQUALITY, ...TEXT_MATCH, ...LIST_MATCH, ...NULLABLE];
  }
}

export function operatorOptions(type?: string): Array<SelectableValue<string>> {
  return operatorsForType(type).map((o) => ({ label: o.label, value: o.value }));
}

const ALL_OPERATORS: OperatorDef[] = [
  ...EQUALITY,
  ...NULLABLE,
  ...COMPARISON,
  ...TEXT_MATCH,
  ...LIST_MATCH,
  ...RANGE,
];

export function operatorArity(op: string): 'none' | 'single' {
  return ALL_OPERATORS.find((o) => o.value === op)?.arity ?? 'single';
}

/**
 * List-style operators take a comma-separated value that the backend compiles
 * into a JSON array (_in/_nin) or a [min,max] range (_between/_nbetween). Their
 * template-variable values are interpolated with `csv` formatting so a
 * multi-value variable expands to comma-separated tokens.
 */
const LIST_OPERATORS = new Set(['in', 'nin', 'between', 'nbetween']);

export function isListOperator(op: string): boolean {
  return LIST_OPERATORS.has(op);
}

// ---------------------------------------------------------------------------
// Template interpolation
// ---------------------------------------------------------------------------

export type ValueInterpolator = (value: string, asList: boolean) => string;

export function interpolateFilterTree(root: FilterGroup, interpolate: ValueInterpolator): FilterGroup {
  const mapNode = (node: FilterNode): FilterNode => {
    if (node.kind === 'group') {
      return { ...node, children: node.children.map(mapNode) };
    }
    if (!node.value) {
      return node;
    }
    return { ...node, value: interpolate(node.value, isListOperator(node.op)) };
  };
  return { ...root, children: root.children.map(mapNode) };
}

// ---------------------------------------------------------------------------
// JSON (de)serialization for persistence on the query
// ---------------------------------------------------------------------------

export function parseFilterTree(raw?: string): FilterGroup {
  if (!raw) {
    return emptyRootGroup();
  }
  try {
    const parsed = JSON.parse(raw);
    if (parsed && parsed.kind === 'group' && Array.isArray(parsed.children)) {
      return parsed as FilterGroup;
    }
  } catch {
    // ignore malformed persisted state
  }
  return emptyRootGroup();
}

export function stringifyFilterTree(root: FilterGroup): string {
  if (root.children.length === 0) {
    return '';
  }
  return JSON.stringify(root);
}
